include/sparrow/layout/run_end_encoded_layout/run_end_encoded_array.hpp

```diff
@@ -0,0 +1,222 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "sparrow/layout/array_wrapper.hpp"
+#include "sparrow/array_factory.hpp"
+#include "sparrow/layout/layout_utils.hpp"
+#include "sparrow/layout/nested_value_types.hpp"
+#include "sparrow/utils/iterator.hpp"
+#include "sparrow/utils/memory.hpp"
+#include "sparrow/utils/nullable.hpp"
+#include "sparrow/layout/dispatch_lib.hpp"
+
+namespace sparrow
+{
+
+    class run_end_encoded_array;
+
+    namespace detail
+    {
+        template<class T>
+        struct get_data_type_from_array;
+
+        template<>
+        struct get_data_type_from_array<sparrow::run_end_encoded_array>
+        {
+            constexpr static sparrow::data_type get()
+            {
+                return sparrow::data_type::RUN_ENCODED;
+            }
+        };
+    }
+
+    // this iteratas over the **actual** values of the run encoded array
+    // Ie nullabes values, not values !!! 
+    template<bool CONST>
+    class run_encoded_array_iterator : public iterator_base<
+        run_encoded_array_iterator<CONST>,
+        array_traits::const_reference,
+        std::forward_iterator_tag,
```

Comment:
Why not mankig it random_access? Is there anything preventing it?

Reply:
random access would be not O(N) but O(logn)